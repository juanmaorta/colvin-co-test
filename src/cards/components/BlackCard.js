/* eslint camelcase: ["off"] */
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Button from '@material-ui/core/Button'
import Card from '@material-ui/core/Card'
import CardActions from '@material-ui/core/CardActions'
import CardContent from '@material-ui/core/CardContent'
import Grid from '@material-ui/core/Grid'
import TextField from '@material-ui/core/TextField'
import Typography from '@material-ui/core/Typography'

import { withStyles } from '@material-ui/core/styles'

import { ENTER_KEY, ESCAPE_KEY } from '../../util/constants'
import styles from '../../styles'

class BlackCard extends Component {
  constructor (props) {
    super(props)

    this.state = {
      isEditing: this.props.card.isEditing || this.props.isEditing,
      text: this.props.card.text,
      undoText: this.props.card.text
    }
  }

  onKeyDown = event => {
    switch (event.keyCode) {
      case ENTER_KEY:
        this.toggleEdit()
        break
      case ESCAPE_KEY:
        this.cancelEdit()
        break
      default:
    }
  }

  cancelEdit = () => {
    this.setState({
      isEditing: !this.state.isEditing,
      text: this.state.undoText
    })
  }

  handleChange = name => event => {
    this.setState({
      [name]: event.target.value
    })
  }

  toggleEdit = () => {
    if (this.state.isEditing) {
      this.props.handleEdit(
        {
          id: this.props.card.id,
          text: this.state.text,
          whiteCards: this.props.card.whiteCards
        }
      )
    }
    this.setState({
      isEditing: !this.state.isEditing,
      undoText: this.state.text
    })
  }

  UNSAFE_componentWillReceiveProps (newProps) {
    if (newProps.card.text !== this.state.text) {
      this.setState({
        text: newProps.card.text
      })
    }
  }

  render () {
    const {
      card,
      classes,
      editable,
      handleDelete
    } = this.props

    return (
      <Grid item sm={6} md={4} lg={3}>
        <Card className={classes.card}>
          <CardContent className={classes.blackCardContent}>
            {!this.state.isEditing &&
              <Typography gutterBottom variant='h5' component='h2' className={classes.blackCardTitle}>{this.state.text}</Typography>
            }
            {this.state.isEditing &&
              <TextField
                id='text'
                label='text'
                className={classes.whiteTextField}
                value={this.state.text}
                onChange={this.handleChange('text')}
                onBlur={this.toggleEdit}
                onKeyDown={this.onKeyDown}
                margin='normal'
              />
            }
          </CardContent>
          {editable &&
            <CardActions className={classes.blackBackGround}>
              <Button size='small' color='primary' href={`cards/${card.id}`}>
              View
              </Button>
              <Button size='small' color='primary' onClick={this.toggleEdit}>
              Edit
              </Button>

              <Button size='small' color='secondary' onClick={() => handleDelete(card.id)}>
              Delete
              </Button>
            </CardActions>
          }
        </Card>
      </Grid>
    )
  }
}

BlackCard.propTypes = {
  classes: PropTypes.object.isRequired,
  card: PropTypes.object.isRequired,
  editable: PropTypes.bool,
  handleDelete: PropTypes.func,
  handleEdit: PropTypes.func
}

BlackCard.defaultProps = {
  editable: false,
  isEditing: false
}

export default withStyles(styles)(BlackCard)
