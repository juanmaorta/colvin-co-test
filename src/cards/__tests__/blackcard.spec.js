/* global expect, describe, it, jest */
import React from 'react'
import { shallow } from 'enzyme'

import BlackCard from '../components/BlackCard'
import * as stubs from './stubs'

describe('BlackCard snapshots', () => {
  it('should match for BlackCard component', () => {
    const wrapper = shallow(
      <BlackCard
        classes={stubs.classes}
        editable
        handleDelete={jest.fn}
        handleEdit={jest.fn}
        card={stubs.card}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })
})
