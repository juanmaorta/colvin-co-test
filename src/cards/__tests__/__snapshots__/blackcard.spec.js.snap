// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlackCard snapshots should match for BlackCard component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithStyles(BlackCard)
    card={
        Object {
            "id": "3c6f4977-c58b-460f-98ad-acd2fae3a2ce",
            "text": "GoldFinger: He's the man, the man with...",
          }
    }
    classes={Object {}}
    editable={true}
    handleDelete={[Function]}
    handleEdit={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "card": Object {
        "id": "3c6f4977-c58b-460f-98ad-acd2fae3a2ce",
        "text": "GoldFinger: He's the man, the man with...",
      },
      "classes": Object {
        "appBar": "BlackCard-appBar-1",
        "blackBackGround": "BlackCard-blackBackGround-5",
        "blackCardContent": "BlackCard-blackCardContent-6",
        "blackCardTitle": "BlackCard-blackCardTitle-3",
        "buttons": "BlackCard-buttons-2",
        "card": "BlackCard-card-4",
        "cardGrid": "BlackCard-cardGrid-12",
        "footer": "BlackCard-footer-7",
        "heroButtons": "BlackCard-heroButtons-10",
        "heroContent": "BlackCard-heroContent-9",
        "heroUnit": "BlackCard-heroUnit-8",
        "layout": "BlackCard-layout-11",
      },
      "editable": true,
      "handleDelete": [Function],
      "handleEdit": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "card": Object {
          "id": "3c6f4977-c58b-460f-98ad-acd2fae3a2ce",
          "text": "GoldFinger: He's the man, the man with...",
        },
        "classes": Object {
          "appBar": "BlackCard-appBar-1",
          "blackBackGround": "BlackCard-blackBackGround-5",
          "blackCardContent": "BlackCard-blackCardContent-6",
          "blackCardTitle": "BlackCard-blackCardTitle-3",
          "buttons": "BlackCard-buttons-2",
          "card": "BlackCard-card-4",
          "cardGrid": "BlackCard-cardGrid-12",
          "footer": "BlackCard-footer-7",
          "heroButtons": "BlackCard-heroButtons-10",
          "heroContent": "BlackCard-heroContent-9",
          "heroUnit": "BlackCard-heroUnit-8",
          "layout": "BlackCard-layout-11",
        },
        "editable": true,
        "handleDelete": [Function],
        "handleEdit": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
